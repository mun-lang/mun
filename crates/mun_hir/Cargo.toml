[package]
name = "mun_hir"
version = "0.1.0"
authors = ["Bas Zalmstra <zalmstra.bas@gmail.com>"]
edition = "2018"

[dependencies]
salsa="0.12"
superslice = "1.0"
mun_syntax={path="../mun_syntax"}
rustc-hash = "1.0"
once_cell = "0.2"
relative-path = "0.4.0"
ena = "0.13"
drop_bomb = "0.1.4"
either = "1.5.3"

[dev-dependencies]
insta = "0.12.0"
