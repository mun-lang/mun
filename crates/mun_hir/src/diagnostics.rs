use crate::{FileId, HirDatabase, Ty};
use mun_syntax::{ast, AstPtr, SyntaxNode, SyntaxNodePtr, TextRange};
use std::{any::Any, fmt};

/// Diagnostic defines hir API for errors and warnings.
///
/// It is used as a `dyn` object, which you can downcast to concrete diagnostics. DiagnosticSink
/// are structured, meaning that they include rich information which can be used by IDE to create
/// fixes.
///
/// Internally, various subsystems of HIR produce diagnostics specific to a subsystem (typically,
/// an `enum`), which are safe to store in salsa but do not include source locations. Such internal
/// diagnostics are transformed into an instance of `Diagnostic` on demand.
pub trait Diagnostic: Any + Send + Sync + fmt::Debug + 'static {
    fn message(&self) -> String;
    fn file(&self) -> FileId;
    fn syntax_node_ptr(&self) -> SyntaxNodePtr;
    fn highlight_range(&self) -> TextRange {
        self.syntax_node_ptr().range()
    }
    fn as_any(&self) -> &(dyn Any + Send + 'static);
}

pub trait AstDiagnostic {
    type AST;
    fn ast(&self, db: &impl HirDatabase) -> Self::AST;
}

impl dyn Diagnostic {
    pub fn syntax_node(&self, db: &impl HirDatabase) -> SyntaxNode {
        let node = db.parse(self.file()).syntax_node();
        self.syntax_node_ptr().to_node(&node)
    }

    pub fn downcast_ref<D: Diagnostic>(&self) -> Option<&D> {
        self.as_any().downcast_ref()
    }
}

type DiagnosticCallback<'a> = Box<dyn FnMut(&dyn Diagnostic) -> Result<(), ()> + 'a>;

pub struct DiagnosticSink<'a> {
    callbacks: Vec<DiagnosticCallback<'a>>,
    default_callback: Box<dyn FnMut(&dyn Diagnostic) + 'a>,
}

impl<'a> DiagnosticSink<'a> {
    pub fn new(cb: impl FnMut(&dyn Diagnostic) + 'a) -> DiagnosticSink<'a> {
        DiagnosticSink {
            callbacks: Vec::new(),
            default_callback: Box::new(cb),
        }
    }

    pub fn on<D: Diagnostic, F: FnMut(&D) + 'a>(mut self, mut cb: F) -> DiagnosticSink<'a> {
        let cb = move |diag: &dyn Diagnostic| match diag.downcast_ref::<D>() {
            Some(d) => {
                cb(d);
                Ok(())
            }
            None => Err(()),
        };
        self.callbacks.push(Box::new(cb));
        self
    }

    pub(crate) fn push(&mut self, d: impl Diagnostic) {
        let d: &dyn Diagnostic = &d;
        for cb in self.callbacks.iter_mut() {
            match cb(d) {
                Ok(()) => return,
                Err(()) => (),
            }
        }
        (self.default_callback)(d)
    }
}

#[derive(Debug)]
pub struct UnresolvedValue {
    pub file: FileId,
    pub expr: SyntaxNodePtr,
}

impl Diagnostic for UnresolvedValue {
    fn message(&self) -> String {
        "undefined value".to_string()
    }

    fn file(&self) -> FileId {
        self.file
    }

    fn syntax_node_ptr(&self) -> SyntaxNodePtr {
        self.expr
    }

    fn as_any(&self) -> &(dyn Any + Send + 'static) {
        self
    }
}

#[derive(Debug)]
pub struct UnresolvedType {
    pub file: FileId,
    pub type_ref: AstPtr<ast::TypeRef>,
}

impl Diagnostic for UnresolvedType {
    fn message(&self) -> String {
        "undefined type".to_string()
    }

    fn file(&self) -> FileId {
        self.file
    }

    fn syntax_node_ptr(&self) -> SyntaxNodePtr {
        self.type_ref.syntax_node_ptr()
    }

    fn as_any(&self) -> &(dyn Any + Send + 'static) {
        self
    }
}

#[derive(Debug)]
pub struct ExpectedFunction {
    pub file: FileId,
    pub expr: SyntaxNodePtr,
    pub found: Ty,
}

impl Diagnostic for ExpectedFunction {
    fn message(&self) -> String {
        "expected function type".to_string()
    }

    fn file(&self) -> FileId {
        self.file
    }

    fn syntax_node_ptr(&self) -> SyntaxNodePtr {
        self.expr
    }

    fn as_any(&self) -> &(dyn Any + Send + 'static) {
        self
    }
}

#[derive(Debug)]
pub struct ParameterCountMismatch {
    pub file: FileId,
    pub expr: SyntaxNodePtr,
    pub expected: usize,
    pub found: usize,
}

impl Diagnostic for ParameterCountMismatch {
    fn message(&self) -> String {
        format!(
            "this function takes {} parameters but {} parameters was supplied",
            self.expected, self.found
        )
    }

    fn file(&self) -> FileId {
        self.file
    }

    fn syntax_node_ptr(&self) -> SyntaxNodePtr {
        self.expr
    }

    fn as_any(&self) -> &(dyn Any + Send + 'static) {
        self
    }
}

#[derive(Debug)]
pub struct MismatchedType {
    pub file: FileId,
    pub expr: SyntaxNodePtr,
    pub expected: Ty,
    pub found: Ty,
}

impl Diagnostic for MismatchedType {
    fn message(&self) -> String {
        "mismatched type".to_string()
    }

    fn file(&self) -> FileId {
        self.file
    }

    fn syntax_node_ptr(&self) -> SyntaxNodePtr {
        self.expr
    }

    fn as_any(&self) -> &(dyn Any + Send + 'static) {
        self
    }
}

#[derive(Debug)]
pub struct IncompatibleBranch {
    pub file: FileId,
    pub if_expr: SyntaxNodePtr,
    pub expected: Ty,
    pub found: Ty,
}

impl Diagnostic for IncompatibleBranch {
    fn message(&self) -> String {
        "mismatched branches".to_string()
    }

    fn file(&self) -> FileId {
        self.file
    }

    fn syntax_node_ptr(&self) -> SyntaxNodePtr {
        self.if_expr
    }

    fn as_any(&self) -> &(dyn Any + Send + 'static) {
        self
    }
}

#[derive(Debug)]
pub struct InvalidLHS {
    /// The file that contains the expressions
    pub file: FileId,

    /// The expression containing the `lhs`
    pub expr: SyntaxNodePtr,

    /// The left-hand side of the expression.
    pub lhs: SyntaxNodePtr,
}

impl Diagnostic for InvalidLHS {
    fn message(&self) -> String {
        "invalid left hand side of expression".to_string()
    }

    fn file(&self) -> FileId {
        self.file
    }

    fn syntax_node_ptr(&self) -> SyntaxNodePtr {
        self.lhs
    }

    fn as_any(&self) -> &(dyn Any + Send + 'static) {
        self
    }
}

#[derive(Debug)]
pub struct MissingElseBranch {
    pub file: FileId,
    pub if_expr: SyntaxNodePtr,
    pub found: Ty,
}

impl Diagnostic for MissingElseBranch {
    fn message(&self) -> String {
        "missing else branch".to_string()
    }

    fn file(&self) -> FileId {
        self.file
    }

    fn syntax_node_ptr(&self) -> SyntaxNodePtr {
        self.if_expr
    }

    fn as_any(&self) -> &(dyn Any + Send + 'static) {
        self
    }
}

#[derive(Debug)]
pub struct CannotApplyBinaryOp {
    pub file: FileId,
    pub expr: SyntaxNodePtr,
    pub lhs: Ty,
    pub rhs: Ty,
}

impl Diagnostic for CannotApplyBinaryOp {
    fn message(&self) -> String {
        "cannot apply binary operator".to_string()
    }

    fn file(&self) -> FileId {
        self.file
    }

    fn syntax_node_ptr(&self) -> SyntaxNodePtr {
        self.expr
    }

    fn as_any(&self) -> &(dyn Any + Send + 'static) {
        self
    }
}

#[derive(Debug)]
pub struct DuplicateDefinition {
    pub file: FileId,
    pub name: String,
    pub first_definition: SyntaxNodePtr,
    pub definition: SyntaxNodePtr,
}

impl Diagnostic for DuplicateDefinition {
    fn message(&self) -> String {
        format!("the name `{}` is defined multiple times", self.name)
    }

    fn file(&self) -> FileId {
        self.file
    }

    fn syntax_node_ptr(&self) -> SyntaxNodePtr {
        self.definition
    }

    fn as_any(&self) -> &(dyn Any + Send + 'static) {
        self
    }
}

#[derive(Debug)]
pub struct ReturnMissingExpression {
    pub file: FileId,
    pub return_expr: SyntaxNodePtr,
}

impl Diagnostic for ReturnMissingExpression {
    fn message(&self) -> String {
        "`return;` in a function whose return type is not `()`".to_owned()
    }

    fn file(&self) -> FileId {
        self.file
    }

    fn syntax_node_ptr(&self) -> SyntaxNodePtr {
        self.return_expr
    }

    fn as_any(&self) -> &(dyn Any + Send + 'static) {
        self
    }
}
