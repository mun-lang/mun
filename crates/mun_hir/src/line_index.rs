use mun_syntax::TextUnit;

use superslice::Ext;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct LineIndex {
    newlines: Vec<TextUnit>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct LineCol {
    pub line: u32,
    pub col: u32,
}

impl LineIndex {
    pub fn new(text: &str) -> LineIndex {
        let mut newlines = vec![0.into()];
        let mut curr_row = 0.into();
        for c in text.chars() {
            curr_row += TextUnit::of_char(c);
            if c == '\n' {
                newlines.push(curr_row);
            }
        }

        LineIndex { newlines }
    }

    pub fn line_col(&self, offset: TextUnit) -> LineCol {
        let line = self.newlines.upper_bound(&offset) - 1;
        let line_start_offset = self.newlines[line];
        let col = offset - line_start_offset;

        LineCol {
            line: line as u32,
            col: col.to_usize() as u32,
        }
    }

    pub fn line_str<'a>(&self, line: u32, text: &'a str) -> Option<&'a str> {
        let start_of_line = self.newlines.get(line as usize)?.to_usize();
        let end_of_line = self
            .newlines
            .get((line + 1) as usize)
            .map(|u| u.to_usize() - 1)
            .unwrap_or_else(|| text.chars().count());
        Some(&text[start_of_line..end_of_line])
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_line_index() {
        let text = "hello\nworld";
        let index = LineIndex::new(text);
        assert_eq!(index.line_col(0.into()), LineCol { line: 0, col: 0 });
        assert_eq!(index.line_col(1.into()), LineCol { line: 0, col: 1 });
        assert_eq!(index.line_col(5.into()), LineCol { line: 0, col: 5 });
        assert_eq!(index.line_col(6.into()), LineCol { line: 1, col: 0 });
        assert_eq!(index.line_col(7.into()), LineCol { line: 1, col: 1 });
    }
}
