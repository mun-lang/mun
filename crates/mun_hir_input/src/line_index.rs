use mun_syntax::TextSize;
use rustc_hash::FxHashMap;

/// A [`LineIndex`] enables efficient mapping between offsets and line/column
/// positions in a text.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct LineIndex {
    /// Offsets from the beginning of each line
    newlines: Vec<TextSize>,

    /// List of non-ASCII characters on each line
    utf16_lines: FxHashMap<u32, Vec<Utf16Char>>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct LineCol {
    /// The line index (zero-based)
    pub line: u32,

    /// The column index when the text is represented as UTF16 text (zero-based)
    pub col_utf16: u32,
}

#[derive(Clone, Debug, Hash, PartialEq, Eq)]
struct Utf16Char {
    /// Start offset of a character inside a line, zero-based
    pub(crate) start: TextSize,

    /// End offset of a character inside a line, zero-based
    pub(crate) end: TextSize,
}

impl Utf16Char {
    /// Returns the length in 8-bit UTF-8 code units.
    fn len(&self) -> TextSize {
        self.end - self.start
    }

    /// Returns the length in 16-bit UTF-16 code units.
    fn len_utf16(&self) -> usize {
        if self.len() == TextSize::from(4) {
            2
        } else {
            1
        }
    }
}

impl LineIndex {
    /// Constructs a new [`LineIndex`] from the given text.
    pub fn new(text: &str) -> LineIndex {
        let mut utf16_lines = FxHashMap::default();
        let mut utf16_chars = Vec::new();

        // Iterate over all the characters in the text and record all the newlines and
        // UTF16 characters.
        let mut newlines = vec![0.into()];
        let mut curr_row = 0.into();
        let mut curr_col = 0.into();
        let mut line = 0;
        for c in text.chars() {
            let c_len = TextSize::of(c);
            curr_row += c_len;
            if c == '\n' {
                newlines.push(curr_row);

                // Save any utf-16 characters seen in the previous line
                if !utf16_chars.is_empty() {
                    utf16_lines.insert(line, utf16_chars);
                    utf16_chars = Vec::new();
                }

                // Prepare for processing the next line
                curr_col = 0.into();
                line += 1;
                continue;
            }

            if !c.is_ascii() {
                utf16_chars.push(Utf16Char {
                    start: curr_col,
                    end: curr_col + c_len,
                });
            }

            curr_col += c_len;
        }

        // Save any utf-16 characters seen in the last line
        if !utf16_chars.is_empty() {
            utf16_lines.insert(line, utf16_chars);
        }

        LineIndex {
            newlines,
            utf16_lines,
        }
    }

    /// Returns the line and column index at the given offset in the text
    pub fn line_col(&self, offset: TextSize) -> LineCol {
        let line = self
            .newlines
            .binary_search_by(|x| {
                if x <= &offset {
                    std::cmp::Ordering::Less
                } else {
                    std::cmp::Ordering::Greater
                }
            })
            .unwrap_or_else(|i| i)
            - 1;
        let line_start_offset = self.newlines[line];
        let col = offset - line_start_offset;

        LineCol {
            line: line as u32,
            col_utf16: self.utf8_to_utf16_col(line as u32, col) as u32,
        }
    }

    /// Returns the offset in the text for the given line and column index
    pub fn offset(&self, line_col: LineCol) -> TextSize {
        let col = self.utf16_to_utf8_col(line_col.line, line_col.col_utf16);
        self.newlines[line_col.line as usize] + col
    }

    /// Retrieves the text between `first_line` and `last_line`, if any.
    pub fn text_part<'a>(
        &self,
        first_line: u32,
        last_line: u32,
        text: &'a str,
        text_len: usize,
    ) -> Option<&'a str> {
        let start_of_part = (*self.newlines.get(first_line as usize)?).into();
        let end_of_part = self
            .newlines
            .get(last_line as usize + 1)
            .map_or(text_len, |u| usize::from(*u) - 1usize);
        Some(&text[start_of_part..end_of_part])
    }

    /// Retrieves the offset to the line corresponding to `line_index`.
    #[inline]
    pub fn line_offset(&self, line_index: u32) -> usize {
        self.newlines[line_index as usize].into()
    }

    /// Given a line and column number for utf16 text convert it to the offset
    /// in utf8 text.
    fn utf16_to_utf8_col(&self, line: u32, mut col: u32) -> TextSize {
        if let Some(utf16_chars) = self.utf16_lines.get(&line) {
            for c in utf16_chars {
                if col > u32::from(c.start) {
                    col += u32::from(c.len()) - c.len_utf16() as u32;
                } else {
                    // From here on, all utf16 characters come *after* the character we are mapping,
                    // so we don't need to take them into account
                    break;
                }
            }
        }

        col.into()
    }

    /// Given a line and column number for utf8 text, convert it to the offset
    /// in utf16 text.
    fn utf8_to_utf16_col(&self, line: u32, col: TextSize) -> usize {
        let mut res: usize = col.into();
        if let Some(utf16_chars) = self.utf16_lines.get(&line) {
            for c in utf16_chars {
                if c.end <= col {
                    res -= usize::from(c.len()) - c.len_utf16();
                } else {
                    // From here on, all utf16 characters come *after* the character we are mapping,
                    // so we don't need to take them into account
                    break;
                }
            }
        }
        res
    }
}
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_line_index() {
        let text = "hello\nworld";
        let index = LineIndex::new(text);
        assert_eq!(
            index.line_col(0.into()),
            LineCol {
                line: 0,
                col_utf16: 0
            }
        );
        assert_eq!(
            index.line_col(1.into()),
            LineCol {
                line: 0,
                col_utf16: 1
            }
        );
        assert_eq!(
            index.line_col(5.into()),
            LineCol {
                line: 0,
                col_utf16: 5
            }
        );
        assert_eq!(
            index.line_col(6.into()),
            LineCol {
                line: 1,
                col_utf16: 0
            }
        );
        assert_eq!(
            index.line_col(7.into()),
            LineCol {
                line: 1,
                col_utf16: 1
            }
        );
    }
    #[test]
    fn test_text_part() {
        let text = "ℱ٥ℜ\n†ěṦτ\nℙน尺קő$ع";
        let text_len = text.len();
        let index = LineIndex::new(text);
        assert_eq!(index.text_part(0, 0, text, text_len), Some("ℱ٥ℜ"));
        assert_eq!(index.text_part(0, 1, text, text_len), Some("ℱ٥ℜ\n†ěṦτ"));
        assert_eq!(
            index.text_part(1, 2, text, text_len),
            Some("†ěṦτ\nℙน尺קő$ع")
        );
        assert_eq!(index.text_part(0, 2, text, text_len), Some(text));
    }
    #[test]
    fn test_text_part_utf16() {
        let text = "a\n❤️\nb";
        let index = LineIndex::new(text);
        let start = index.offset(LineCol {
            line: 1,
            col_utf16: 0,
        });
        let end = index.offset(LineCol {
            line: 1,
            col_utf16: 1,
        });
        assert_eq!(
            index.text_part(1, 1, text, (end - start).into()),
            Some("❤️")
        );
    }

    #[test]
    fn test_line_offset() {
        let text = "for\ntest\npurpose";
        let index = LineIndex::new(text);
        assert_eq!(index.line_offset(0), 0);
        assert_eq!(index.line_offset(1), 4);
        assert_eq!(index.line_offset(2), 9);
    }
}
