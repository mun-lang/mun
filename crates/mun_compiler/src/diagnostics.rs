use mun_hir::diagnostics::{Diagnostic as HirDiagnostic, DiagnosticSink};
use mun_hir::{FileId, HirDatabase, HirDisplay, Module};
use mun_syntax::{ast, AstNode, SyntaxKind};
use std::cell::RefCell;

mod emit;

pub use emit::Emit;
use mun_errors::{Diagnostic, Level};

/// Constructs diagnostics message for the given file.
pub fn diagnostics(db: &impl HirDatabase, file_id: FileId) -> Vec<Diagnostic> {
    let parse = db.parse(file_id);
    let mut result = Vec::new();

    result.extend(parse.errors().iter().map(|err| Diagnostic {
        level: Level::Error,
        loc: err.location(),
        message: format!("Syntax Error: {}", err),
    }));

    let result = RefCell::new(result);
    let mut sink = DiagnosticSink::new(|d| {
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: d.highlight_range().into(),
            message: d.message(),
        });
    })
    .on::<mun_hir::diagnostics::UnresolvedValue, _>(|d| {
        let text = d.expr.to_node(&parse.tree().syntax()).text().to_string();
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: d.highlight_range().into(),
            message: format!("could not find value `{}` in this scope", text),
        });
    })
    .on::<mun_hir::diagnostics::UnresolvedType, _>(|d| {
        let text = d
            .type_ref
            .to_node(&parse.tree().syntax())
            .syntax()
            .text()
            .to_string();
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: d.highlight_range().into(),
            message: format!("could not find type `{}` in this scope", text),
        });
    })
    .on::<mun_hir::diagnostics::ExpectedFunction, _>(|d| {
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: d.highlight_range().into(),
            message: format!("expected function, found `{}`", d.found.display(db)),
        });
    })
    .on::<mun_hir::diagnostics::MismatchedType, _>(|d| {
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: d.highlight_range().into(),
            message: format!(
                "expected `{}`, found `{}`",
                d.expected.display(db),
                d.found.display(db)
            ),
        });
    })
    .on::<mun_hir::diagnostics::DuplicateDefinition, _>(|d| {
        result.borrow_mut().push(Diagnostic {
            level: Level::Error,
            loc: match d.definition.kind() {
                SyntaxKind::FUNCTION_DEF => {
                    ast::FunctionDef::cast(d.definition.to_node(&parse.tree().syntax()))
                        .map(|f| f.signature_range())
                        .unwrap_or_else(|| d.highlight_range())
                        .into()
                }
                _ => d.highlight_range().into(),
            },
            message: d.message(),
        });
    });

    Module::from(file_id).diagnostics(db, &mut sink);

    drop(sink);
    result.into_inner()
}
