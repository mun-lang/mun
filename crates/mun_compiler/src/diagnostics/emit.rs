use mun_errors::Diagnostic;
use mun_hir::{FileId, SourceDatabase};
use std::io;
use termcolor::{Color, ColorSpec, WriteColor};

pub trait Emit {
    fn emit(
        &self,
        writer: &mut impl WriteColor,
        db: &impl SourceDatabase,
        file_id: FileId,
    ) -> io::Result<()>;
}

impl Emit for Diagnostic {
    fn emit(
        &self,
        writer: &mut impl WriteColor,
        db: &impl SourceDatabase,
        file_id: FileId,
    ) -> io::Result<()> {
        let line_index = db.line_index(file_id);
        let text = db.file_text(file_id).to_string();
        let path = db.file_relative_path(file_id);
        let line_col = line_index.line_col(self.loc.offset());
        let line_col_end = line_index.line_col(self.loc.end_offset());

        let header = ColorSpec::new()
            .set_fg(Some(Color::White))
            .set_bold(true)
            .set_intense(true)
            .clone();
        let error = header.clone().set_fg(Some(Color::Red)).clone();
        let snippet_gutter = ColorSpec::new()
            .set_fg(Some(Color::Cyan))
            .set_bold(true)
            .set_intense(true)
            .clone();
        let snippet_text = ColorSpec::new();

        // Write severity name
        writer.set_color(&error)?;
        write!(writer, "error")?;

        // Write diagnostics message
        writer.set_color(&header)?;
        writeln!(writer, ": {}", self.message)?;

        if let Some(snippet) = line_index.line_str(line_col.line, &text) {
            // Determine gutter offset
            let line_str = format!("{}", line_col.line + 1);
            let gutter_indent = " ".to_string().repeat(line_str.len());

            writer.set_color(&snippet_gutter)?;
            write!(writer, "{}-->", gutter_indent)?;
            writer.set_color(&snippet_text)?;
            writeln!(
                writer,
                " {}:{}:{}",
                path.as_str(),
                line_col.line + 1,
                line_col.col
            )?;

            // Snippet
            writer.set_color(&snippet_gutter)?;
            writeln!(writer, "{} |", gutter_indent)?;
            write!(writer, "{} | ", line_str)?;
            writer.set_color(&snippet_text)?;
            writeln!(writer, "{}", snippet)?;

            writer.set_color(&snippet_gutter)?;
            write!(writer, "{} |", gutter_indent)?;
            writer.set_color(&error)?;

            if line_col.line == line_col_end.line {
                // single-line diagnostics
                writeln!(
                    writer,
                    " {}{}",
                    " ".to_string().repeat(line_col.col as usize),
                    "^".to_string()
                        .repeat((line_col_end.col - line_col.col) as usize)
                )?;
            }
        }

        //        // Write the start location
        //        println!("  {} {}:{}:{}",
        //            "-->".cyan(),
        //            path.as_str(),
        //            line_col.line + 1,
        //            line_col.col);

        writer.reset()?;
        Ok(())
    }
}
