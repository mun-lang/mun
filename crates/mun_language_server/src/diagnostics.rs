use std::cell::RefCell;

use mun_diagnostics::DiagnosticForWith;
use mun_hir::{AstDatabase, InFile, Module};
use mun_hir_input::{FileId, ModuleId, PackageId, SourceDatabase};
use mun_syntax::{Location, TextRange};

use crate::db::AnalysisDatabase;

#[derive(Debug)]
pub struct SourceAnnotation {
    pub message: String,
    pub range: InFile<TextRange>,
}

#[derive(Debug)]
pub struct Diagnostic {
    pub message: String,
    pub range: TextRange,
    pub additional_annotations: Vec<SourceAnnotation>,
    // pub fix: Option<SourceChange>,
    // pub severity: Severity,
}

/// Converts a location to a a range for use in diagnostics
fn location_to_range(location: Location) -> TextRange {
    match location {
        Location::Offset(offset) => TextRange::at(offset, 1.into()),
        Location::Range(range) => range,
    }
}

/// Computes all the diagnostics for the specified file.
pub(crate) fn diagnostics(db: &AnalysisDatabase, file_id: FileId) -> Vec<Diagnostic> {
    let mut result = Vec::new();

    // Add all syntax errors
    let parse = db.parse(file_id);
    result.extend(parse.errors().iter().map(|err| Diagnostic {
        message: format!("parse error: {err}"),
        range: location_to_range(err.location()),
        additional_annotations: vec![],
    }));

    // Add all HIR diagnostics
    let result = RefCell::new(result);
    let mut sink = mun_hir::diagnostics::DiagnosticSink::new(|d| {
        result.borrow_mut().push(d.with_diagnostic(db, |d| {
            Diagnostic {
                message: format!("{}\n{}", d.title(), d.footer().join("\n"))
                    .trim()
                    .to_owned(),
                range: d.range(),
                additional_annotations: d
                    .secondary_annotations()
                    .into_iter()
                    .map(|annotation| SourceAnnotation {
                        message: annotation.message,
                        range: annotation.range,
                    })
                    .collect(),
            }
        }));
    });

    let package_id = PackageId(0);
    let module_tree = db.module_tree(package_id);
    if let Some(local_id) = module_tree.module_for_file(file_id) {
        let module_id = ModuleId {
            package: package_id,
            local_id,
        };
        Module::from(module_id).diagnostics(db, &mut sink);
    }
    drop(sink);

    // Returns the result
    result.into_inner()
}
