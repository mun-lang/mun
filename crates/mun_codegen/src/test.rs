use crate::{mock::MockDatabase, IrDatabase};
use mun_hir::diagnostics::DiagnosticSink;
use mun_hir::line_index::LineIndex;
use mun_hir::Module;
use mun_hir::SourceDatabase;
use std::cell::RefCell;
use std::sync::Arc;

#[test]
fn function() {
    test_snapshot(
        r#"
    fn main() {
    }
    "#,
    );
}

#[test]
fn return_type() {
    test_snapshot(
        r#"
    fn main():int {
      0
    }
    "#,
    );
}

#[test]
fn function_arguments() {
    test_snapshot(
        r#"
    fn main(a:int):int {
      a
    }
    "#,
    );
}

#[test]
fn binary_expressions() {
    test_snapshot(
        r#"
    fn add(a:int, b:int):int {
      a+b
    }

    fn subtract(a:int, b:int):int {
      a-b
    }

    fn multiply(a:int, b:int):int {
      a*b
    }
    "#,
    );
}

#[test]
fn let_statement() {
    test_snapshot(
        r#"
    fn main(a:int):int {
      let b = a+1
      b
    }
    "#,
    );
}

#[test]
fn invalid_binary_ops() {
    test_snapshot(
        r#"
    fn main() {
      let a = 3+3.0;
      let b = 3.0+3;
    }
    "#,
    );
}

#[test]
fn update_operators() {
    test_snapshot(
        r#"
    fn add(a:int, b:int):int {
      let result = a
      result += b
      result
    }

    fn subtract(a:int, b:int):int {
      let result = a
      result -= b
      result
    }

    fn multiply(a:int, b:int):int {
      let result = a
      result *= b
      result
    }

    fn divide(a:int, b:int):int {
      let result = a
      result /= b
      result
    }
    "#,
    );
}

#[test]
fn update_parameter() {
    test_snapshot(
        r#"
    fn add_three(a:int):int {
      a += 3;
      a
    }
    "#,
    );
}

#[test]
fn function_calls() {
    test_snapshot(
        r#"
    fn add_impl(a:int, b:int):int {
        a+b
    }

    fn add(a:int, b:int):int {
      add_impl(a,b)
    }

    fn test():int {
      add(4,5)
      add_impl(4,5)
      add(4,5)
    }
    "#,
    );
}

#[test]
fn equality_operands() {
    test_snapshot(
        r#"
    fn equals(a:int, b:int):bool                { a == b }
    fn not_equals(a:int, b:int):bool            { a != b }
    fn less(a:int, b:int):bool                  { a < b }
    fn less_equal(a:int, b:int):bool            { a <= b }
    fn greater(a:int, b:int):bool               { a > b }
    fn greater_equal(a:int, b:int):bool         { a >= b }
    fn equalsf(a:float, b:float):bool           { a == b }
    fn not_equalsf(a:float, b:float):bool       { a != b }
    fn lessf(a:float, b:float):bool             { a < b }
    fn less_equalf(a:float, b:float):bool       { a <= b }
    fn greaterf(a:float, b:float):bool          { a > b }
    fn greater_equalf(a:float, b:float):bool    { a >= b }
    "#,
    );
}

#[test]
fn if_statement() {
    test_snapshot(
        r#"
    fn foo(a:int):int {
        let b = if a > 3 {
            let c = if a > 4 {
                a+1
            } else {
                a+3
            }
            c
        } else {
            a-1
        }
        b
    }
    "#,
    )
}

#[test]
fn void_return() {
    test_snapshot(
        r#"
    fn bar() {
        let a = 3;
    }
    fn foo(a:int) {
        let c = bar()
    }
    "#,
    )
}

#[test]
fn fibonacci() {
    test_snapshot(
        r#"
    fn fibonacci(n:int):int {
        if n <= 1 {
            n
        } else {
            fibonacci(n-1) + fibonacci(n-2)
        }
    }
    "#,
    )
}

#[test]
fn fibonacci_loop() {
    test_snapshot(
        r#"
    fn fibonacci(n:int):int {
        let a = 0;
        let b = 1;
        let i = 1;
        loop {
            if i > n {
                return a
            }
            let sum = a + b;
            a = b;
            b = sum;
            i += 1;
        }
    }
    "#,
    )
}

#[test]
fn shadowing() {
    test_snapshot(
        r#"
    fn foo(a:int):int {
        let a = a+1;
        {
            let a = a+2;
        }
        a+3
    }

    fn bar(a:int):int {
        let a = a+1;
        let a = {
            let a = a+2;
            a
        }
        a+3
    }
    "#,
    );
}

#[test]
fn return_expr() {
    test_snapshot(
        r#"
    fn main():int {
        return 5;
        let a = 3; // Nothing regarding this statement should be in the IR
    }
    "#,
    );
}

#[test]
fn conditional_return_expr() {
    test_snapshot(
        r#"
    fn main(a:int):int {
        if a > 4 {
            return a;
        }
        a - 1
    }
    "#,
    );
}

#[test]
fn never_conditional_return_expr() {
    test_snapshot(
        r#"
    fn main(a:int):int {
        if a > 4 {
            return a;
        } else {
            return a - 1;
        }
    }
    "#,
    );
}

#[test]
fn true_is_true() {
    test_snapshot(
        r#"
    fn test_true():bool {
        true
    }

    fn test_false():bool {
        false
    }"#,
    );
}

#[test]
fn loop_expr() {
    test_snapshot(
        r#"
    fn foo() {
        loop {}
    }
    "#,
    )
}

#[test]
fn loop_break_expr() {
    test_snapshot(
        r#"
    fn foo(n:int):int {
        loop {
            if n > 5 {
                break n;
            }
            if n > 10 {
                break 10;
            }
            n += 1;
        }
    }
    "#,
    )
}

#[test]
fn while_expr() {
    test_snapshot(
        r#"
    fn foo(n:int) {
        while n<3 {
            n += 1;
        };

        // This will be completely optimized out
        while n<4 {
            break;
        };
    }
    "#,
    )
}

fn test_snapshot(text: &str) {
    let text = text.trim().replace("\n    ", "\n");

    let (db, file_id) = MockDatabase::with_single_file(&text);

    let line_index: Arc<LineIndex> = db.line_index(file_id);
    let messages = RefCell::new(Vec::new());
    let mut sink = DiagnosticSink::new(|diag| {
        let line_col = line_index.line_col(diag.highlight_range().start());
        messages.borrow_mut().push(format!(
            "error {}:{}: {}",
            line_col.line + 1,
            line_col.col + 1,
            diag.message()
        ));
    });
    Module::from(file_id).diagnostics(&db, &mut sink);
    drop(sink);
    let messages = messages.into_inner();

    let name = if !messages.is_empty() {
        messages.join("\n")
    } else {
        format!(
            "{}",
            db.module_ir(file_id)
                .llvm_module
                .print_to_string()
                .to_string()
        )
    };
    insta::assert_snapshot!(insta::_macro_support::AutoName, name, &text);
}
