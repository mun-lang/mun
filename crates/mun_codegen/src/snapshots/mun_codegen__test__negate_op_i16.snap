---
source: crates/mun_codegen/src/test.rs
assertion_line: 1045
expression: "\n    pub fn negate(a: i16) -> i16 { -a }\n                        "
---
; == FILE IR (mod) =====================================
; ModuleID = 'mod'
source_filename = "mod"

@global_type_lookup_table = external global [2 x i64*]

define i16 @negate(i16 %0) {
body:
  %neg = sub i16 0, %0
  ret i16 %neg
}

; == GROUP IR (mod) ====================================
; ModuleID = 'group_name'
source_filename = "group_name"

@global_type_lookup_table = global [2 x i64*] zeroinitializer

